pinot-core/src/main/java/com/linkedin/pinot/core/startree/v2/builder/BaseSingleTreeBuilder.java

```diff
@@ -0,0 +1,477 @@
+      return new Record(dimensions, metrics);
+    } else {
+      for (int i = 0; i < _numMetrics; i++) {
+        record._metrics[i] = _aggregationFunctions[i].applyRawValue(record._metrics[i], segmentRecord._metrics[i]);
+      }
+      return record;
+    }
+  }
+
+  /**
+   * Applies a star-tree record (aggregated record) to the current record.
+   *
+   * @param record Current record
+   * @param starTreeRecord Star-tree record
+   * @return Aggregated record
+   */
+  @SuppressWarnings("unchecked")
+  Record applyStarTreeRecord(@Nullable Record record, Record starTreeRecord) {
+    if (record == null) {
+      int[] dimensions = Arrays.copyOf(starTreeRecord._dimensions, _numDimensions);
+      Object[] metrics = new Object[_numMetrics];
+      for (int i = 0; i < _numMetrics; i++) {
+        metrics[i] = _aggregationFunctions[i].cloneAggregatedValue(starTreeRecord._metrics[i]);
+      }
+      return new Record(dimensions, metrics);
+    } else {
+      for (int i = 0; i < _numMetrics; i++) {
+        record._metrics[i] =
+            _aggregationFunctions[i].applyAggregatedValue(record._metrics[i], starTreeRecord._metrics[i]);
+      }
+      return record;
+    }
+  }
+
+  @Override
+  public void build() throws Exception {
+    long startTime = System.currentTimeMillis();
+    LOGGER.info("Start building star-trees with config: {}", _builderConfig);
+
+    int numSegmentRecords = _segment.getSegmentMetadata().getTotalRawDocs();
+    Iterator<Record> recordIterator = sortAndAggregateSegmentRecords(numSegmentRecords);
+    while (recordIterator.hasNext()) {
+      appendToStarTree(recordIterator.next());
+    }
+    int numStarTreeRecords = _numDocs;
+    LOGGER.info("Generated {} star-tree records from {} segment records", numSegmentRecords, numStarTreeRecords);
+
+    int numRecordsUnderStarNode = _numDocs - numStarTreeRecords;
+    constructStarTree(_rootNode, 0, _numDocs);
+    LOGGER.info("Finish constructing star-tree, got {} tree nodes and {} records under star-node", _numNodes,
+        numRecordsUnderStarNode);
+
+    createAggregatedDocs(_rootNode);
+    LOGGER.info("Finish creating aggregated documents, got {} aggregated records", _numDocs - numRecordsUnderStarNode);
+
+    createForwardIndexes();
+    StarTreeBuilderUtils.serializeTree(new File(_outputDir, STAR_TREE_INDEX_FILE_NAME), _rootNode,
+        _dimensionsSplitOrder, _numNodes);
+    writeMetadata();
+
+    LOGGER.info("Finish building star-tree in {}ms", System.currentTimeMillis() - startTime);
+  }
+
+  private void appendToStarTree(Record record) throws IOException {
+    appendRecord(record);
+    _numDocs++;
+  }
+
+  private TreeNode getNewNode() {
+    _numNodes++;
+    return new TreeNode();
+  }
+
+  private void constructStarTree(TreeNode node, int startDocId, int endDocId) throws IOException {
+    int childDimensionId = node._dimensionId + 1;
+    if (childDimensionId == _numDimensions) {
+      return;
+    }
+
+    // Construct all children nodes
```

Comment:
May be factor out child node and star node creations into separate methods, to make the top level method concise?

Reply:
Done.